Give parrots the ability to repeat learned phrases

arimah opened the pull request:
## About the PR
This PR gives parrots (MobParrot) the ability to repeat overheard phrases or parts of phrases. It accomplishes this by doing the following:

* Adds a new component, `ParrotSpeechComponent`, which defines various parameters and keeps track of learned phrases and timings.
* Adds a new system, `ParrotSpeechSystem`, which listens for nearby utterances and makes the birds occasionally say one of their learned phrases.
* Updates `MobParrot` to include a ParrotSpeech component, as well as ActiveListener (otherwise it won't receive ListenEvents).

## Why / Balance
There is a guaranteed parrot on the Pirate's Cove, as well as one on the Crescent for whatever reason. In an attempt to inject some more life into them, why not have parrots, well, parrot the words they hear?

The various timing parameters probably have to be tweaked somewhat. I don't want parrots to talk non-stop, as it quickly gets tiring, but it also shouldn't be so rare that you never encounter a repeated phrase at all.

* With a learn chance of 20%, after hearing 10 messages, there's an ~89% chance the parrot will have learned at least one phrase, which seems acceptable.
* A parroted phrase every 2–5 minutes results in at *most* ~183 utterances in a 6-hour shift.
* With parrots being relatively rare (can't buy them at cargo), the feature is unlikely to be abused. Parrots also have a cap to the number of phrases they can learn, as well as the length (in number of words) of each phrase. Parrots do not forget old phrases over time, however.
* Phrases are randomly selected sequences of words from an overheard message. Parrots will prefer longer phrases, but can end up learning single words too. Phrases can span across sentences (e.g., "that Pioneer. Seems very") and otherwise not make sense without context (e.g., "all of the"). This is not only acceptable but desirable: the parrot has no idea what it's saying.
* The accent of the original message is retained. Parrots will faithfully recreate the hisssing ssspeech of lizardsss, the buzzz of moths, wotevah da goblins are sayin', and so on. Also intended and fun.
* Parrot messages do not show up in the chat log, to reduce spam, same as advertisements from vending machines and the like.
* Parrot messages _also_ do not show up in admin logs. If a parrot is found to be flying around Frontier saying bannably offensive things, the admin will have to find the player that taught the parrot that phrase. It is fairly trivial to add an admin log entry like "Parrot (id) learned the phrase (phrase) from (source)", if desired.
* Parrots do *not* learn from each other's utterances. If they did, they'd pick up on the various "RAWWK!" and "SQUAWK!" things the parrot accent adds, which would inevitably pollute the list of learned phrases, possibly with a fun snowball effect.
* When anything with a ParrotSpeechComponent has an active mind, the random chatter is paused.

I have no current intention to upstream this.

## How to test
1. Spawn a parrot.
2. Update its ParrotSpeechComponent to reduce the delay between messages, and set the LearnChance to 1.
3. Make sure you are a living mob near the parrot, not a ghost.
4. Speak.
5. Wait for the parrot to repeat parts of what you've said.
6. Spawn more parrots, play around. :)
7. Ensure that the parrots *don't* repeat each other's utterances.
8. Put vending machines around a parrot. Wait for them to talk. Make sure the parrot does not repeat advertisements.
9. Kill the parrots. They should cease their mindless chatter.

## Media
(TODO)

- [ ] I have added screenshots/videos to this PR showcasing its changes ingame, **or** this PR does not require an ingame showcase

## Breaking changes
No.

**Changelog**
:cl:
- add: Parrots will now occasionally repeat some of the things you say near them. Better watch your words when you plan your next act of piracy!

Cheackraze commented:
is discord integration scoped? I know several SS13 servers had a "Polly the Parrot" that would relay its parrotings to discord general channel

arimah commented:
> is discord integration scoped? I know several SS13 servers had a "Polly the Parrot" that would relay its parrotings to discord general channel

I had no idea! I could look into how that's done for other things. It's not like it would be impossible to add, considering there's a Discord relay for other things. However, if posted in real time, there's a decently high chance it would leak current round info, especially pirate plans. If that happens commonly, pirates will probably just end up killing their parrot.

Cheackraze commented:
The one I've seen is on Russ_Money's community server/discord, and I can say thats a relatively safe environment to check out to see how it works in their discord the "Honk Squad". Its where I've seen it, tho admittedly I havent interacted with the system personally to give any more concrete direction. Other than seeing the relay say some funny stuff in general chat

whatston3 commented on Content.Server/_NF/Speech/Components/ParrotSpeechComponent.cs:
```diff
@@ -0,0 +1,40 @@
+    [DataField]
+    public EntityWhitelist Blacklist { get; private set; } = new();
+
+    [DataField]
```
Shouldn't need this defined in YAML.
```suggestion
    [ViewVariables]
```

whatston3 reviewed (commented):
Looks alright, bounds in OnListen are good, works well.  The parrot can't talk when it's dead, so that's good.

I've got a suggestion to merge this with the wearable shoulder parrot on
[whatston3:parrot-speech-suggestions](https://github.com/arimah/frontier-station-14/compare/parrot-speech...whatston3:frontier-station-14:parrot-speech-suggestions?expand=1).  It's a bit goofy at the moment, since you can still wear the parrot when it's dead, and throw it a few centimeters from your body, but being able to both wear the bird and hear it talk is nice.

arimah commented:
Good catch on dead parrots being unable to talk. The system will keep trying to generate messages, but they'll be blocked by the dead mobstate. I will add it to the steps for testing, to make sure we don't forget to detect instances of necromantic parrot speech.

arimah commented:
@whatston3 I've incorporated your changes, everything seems to work fine. The cage thing is a little bit cursed, but also funny and works perfectly. The parrot that spawns on the Cove is not an item, so can't be picked up and worn, but maybe that's okay.

I've also rebased my branch against the most recent master. I'll be taking this out of draft now. :)

arimah commented:
I'm dumb. You can absolutely pick up the default parrot. I forgor you have to right-click them to do that. Ignore that part of my comment.

whatston3 commented:
> I'm dumb. You can absolutely pick up the default parrot. I forgor you have to right-click them to do that. Ignore that part of my comment.

Not dumb, it's unintuitive.  I think it should probably be done on alt+click, but I'm unsure if that can be set up easily.
